sensing/autoware_pointcloud_preprocessor/src/concatenate_data/concatenate_and_time_sync_node.cpp

```
@@ -0,0 +1,500 @@
+  params_.is_motion_compensated = declare_parameter<bool>("is_motion_compensated");
+  params_.publish_synchronized_pointcloud =
+    declare_parameter<bool>("publish_synchronized_pointcloud");
+  params_.keep_input_frame_in_synchronized_pointcloud =
+    declare_parameter<bool>("keep_input_frame_in_synchronized_pointcloud");
+  params_.publish_previous_but_late_pointcloud =
+    declare_parameter<bool>("publish_previous_but_late_pointcloud");
+  params_.synchronized_pointcloud_postfix =
+    declare_parameter<std::string>("synchronized_pointcloud_postfix");
+  params_.input_twist_topic_type = declare_parameter<std::string>("input_twist_topic_type");
+  params_.input_topics = declare_parameter<std::vector<std::string>>("input_topics");
+  params_.output_frame = declare_parameter<std::string>("output_frame");
+  params_.lidar_timestamp_offsets =
+    declare_parameter<std::vector<double>>("lidar_timestamp_offsets");
+  params_.lidar_timestamp_noise_window =
+    declare_parameter<std::vector<double>>("lidar_timestamp_noise_window");
+
+  if (params_.input_topics.empty()) {
+    RCLCPP_ERROR(get_logger(), "Need a 'input_topics' parameter to be set before continuing!");
+    return;
+  } else if (params_.input_topics.size() == 1) {
+    RCLCPP_ERROR(get_logger(), "Only one topic given. Need at least two topics to continue.");
+    return;
+  }
+
+  if (params_.output_frame.empty()) {
+    RCLCPP_ERROR(get_logger(), "Need an 'output_frame' parameter to be set before continuing!");
+    return;
+  }
+  if (params_.lidar_timestamp_offsets.size() != params_.input_topics.size()) {
+    RCLCPP_ERROR(
+      get_logger(), "The number of topics does not match the number of timestamp offsets");
+    return;
+  }
+  if (params_.lidar_timestamp_noise_window.size() != params_.input_topics.size()) {
+    RCLCPP_ERROR(
+      get_logger(), "The number of topics does not match the number of timestamp noise windwo");
+    return;
+  }
```

Early `return`s in a constructor are not a good way of handling errors:
The caller will have no idea that something went wrong. Throwing an exception is better. In that case, the error message can be moved into the exception.

---

Thanks! fixed in https://github.com/autowarefoundation/autoware.universe/pull/8300/commits/b0c8a7c479b189cd3ab0772598bb0bac9f75ec06
I will keep the rclcpp error for logging and add a boolean variable to determine whether I should throw a runtime error.

---

I don't think this is necessary (or good for the user):
If the user screwed up the configuration, it's in my opinion the best and safest way to stop the program.
Otherwise, you would have to support and test all the unsupported/nonsensical cases. Especially for the size mismatch checks, not throwing but continuing to run is a sure way to run into out-of-bounds errors during execution later.

For autonomous driving, we have to ensure that _during the drive_ nothing crashes/fails. For misconfigurations, that means failing _before the dive_ starts.

My suggestion would be to just replace the logging with throwing and making the logged text the exception message.

---

Understood, thanks!
Fixed them in https://github.com/autowarefoundation/autoware.universe/pull/8300/commits/92d69a4c6a4bb13c0ef2aa3d5e593fbda283c247